Mathlib/Algebra/Group/ForwardDiff.lean

```diff
@@ -191,5 +204,198 @@ lemma fwdDiff_addChar_eq {M R : Type*} [AddCommMonoid M] [Ring R]
+    fwdDiff_iter_finset_sum, ← smul_eq_mul, ← Pi.smul_def, fwdDiff_iter_const_smul, Pi.smul_apply]
+  rw [sum_apply, sum_range_succ, sum_eq_zero (fun i hi ↦ ?_), zero_add,
+    fwdDiff_iter_eq_factorial, leadingCoeff, Pi.smul_apply]
+  rw [fwdDiff_iter_pow_eq_zero_of_lt (mem_range.mp hi), smul_zero, Pi.zero_apply]
+
+theorem Polynomial.fwdDiff_iter_eq_zero_of_degree_lt {P : R[X]} {n : ℕ} (hP : P.natDegree < n) :
+    Δ_[1]^[n] P.eval = 0 := funext fun x ↦ by
+  obtain ⟨j, rfl⟩ := Nat.exists_eq_add_of_lt hP
+  rw [add_assoc, add_comm, Function.iterate_add_apply, Function.iterate_succ_apply,
+    P.fwdDiff_iter_degree_eq_factorial, Pi.smul_def]
+  simp [fwdDiff_iter_eq_sum_shift]
+
+theorem Polynomial.fwdDiff_iter_degree_add_one_eq_zero (P : R[X]) :
+    Δ_[1]^[P.natDegree + 1] P.eval = 0 := by
+    let hP : P.natDegree < P.natDegree + 1 := Nat.lt_succ_self P.natDegree
+    exact Polynomial.fwdDiff_iter_eq_zero_of_degree_lt hP
+
+/--
+The `(n+1)`-th forward difference of a polynomial of degree at most `n` is zero.
+A polynomial `P(x) = ∑_{k=0..n} aₖ xᵏ` has `Δ^[n+1] P = 0`.
+-/
+theorem fwdDiff_iter_succ_sum_eq_zero {n : ℕ} (P : ℕ → R) :
+    Δ_[1]^[n + 1] (fun (r : R) => ∑ k ∈ range (n + 1), P k • r ^ k) = 0 := by
+  induction n with
+  | zero => aesop
+  | succ n IH =>
+    have : Δ_[1]^[n + 1] (fun (r : R) => ∑ k ∈ range (n + 1), P k • r ^ k) =
+      ∑ k ∈ range (n + 1), Δ_[1]^[n + 1] (P k • fun (r : R) => r ^ k) := by
+      simp only [fwdDiff_iter_const_smul]
+      have : (fun (r : R) => ∑ k ∈ range (n + 1), P k • r ^ k) = (∑ k ∈ range (n + 1),
+        (fun (r : R) => P k • r ^ k)) := by
+        ext r; simp only [sum_apply]
+      rw [this]
+      simp only [fwdDiff_iter_finset_sum]
+      congr 1; ext k
+      rw [← fwdDiff_iter_const_smul]
+      have : (fun (r : R) => P k • r ^ k) = P k • (fun (r : R) => r ^ k) := by
+        ext r
+        rw [Pi.smul_apply]
+      rw [this]
+    rw [this] at IH
+    rw [iterate_succ_apply]
+    have : Δ_[1] (fun (r : R) => ∑ k ∈ range (n + 1 + 1), P k • r ^ k) =
+      ∑ k ∈ range (n + 1 + 1), P k • (fun (r : R) ↦ (r + 1) ^ k) +
+        (-1 • ∑ k ∈ range (n + 1 + 1), P k • fun (r : R) ↦ r ^ k) := by
+      unfold fwdDiff
+      ext X
+      simp only [smul_eq_mul, Int.reduceNeg, neg_smul, one_smul, Pi.add_apply, sum_apply,
+        Pi.smul_apply, Pi.neg_apply, sub_eq_add_neg]
+    rw [this, fwdDiff_iter_add, fwdDiff_iter_const_smul]
+    simp only [fwdDiff_iter_finset_sum, fwdDiff_iter_const_smul, Int.reduceNeg, neg_smul, one_smul]
+    rw [← sub_eq_add_neg]
+    nth_rw 2 [sum_range_succ]
+    conv_lhs => enter [2, 1, 2]; ext k; rw [← fwdDiff_iter_const_smul]
+    rw [IH]
+    simp only [zero_add]
+    rw [fwdDiff_iter_eq_factorial, sum_range_succ, fwdDiff_iter_eq_factorial']
+    simp only [add_sub_cancel_right]
+    exact sum_eq_zero fun i hi ↦ by
+      rw [fwdDiff_iter_pow_eq_zero_of_lt' (by have := mem_range.1 hi; omega), smul_zero]
```

Comment:
Here's a much shorter proof:
```suggestion
/--
The `n`-th forward difference of a polynomial of degree `< n` is zero (formulated using explicit
sums over `range n`.
-/
theorem fwdDiff_iter_sum_mul_pow_eq_zero {n : ℕ} (P : ℕ → R) :
    Δ_[1]^[n] (fun r : R ↦ ∑ k ∈ range n, P k * r ^ k) = 0 := by
  simp_rw [← sum_apply _ _ (fun i x ↦ P i * x ^ i), fwdDiff_iter_finset_sum, sum_fn, ← smul_eq_mul,
    ← Pi.smul_def, fwdDiff_iter_const_smul, ← sum_fn]
  exact sum_eq_zero fun i hi ↦ smul_eq_zero_of_right _ <| fwdDiff_iter_pow_eq_zero_of_lt
    <| mem_range.mp hi
```

If you make this change you can also get rid of the lemmas `fwdDiff_iter_eq_factorial'` and `fwdDiff_iter_pow_eq_zero_of_lt' ` (with the prime `'`) above, since they were only needed in this proof.